File: Data/Engine/Containers/webui-frontend/data/web-interface/src/app/utils/inputValidation.js
const DEFAULT_MAX_LENGTH = 4096;
const IDENTIFIER_MAX_LENGTH = 256;
const PATH_MAX_LENGTH = 4096;
const SECRET_MAX_LENGTH = 16 * 1024 * 1024;

const FIELD_CLASS = {
  PLAIN_SINGLE_LINE: "plain_single_line",
  PLAIN_MULTILINE: "plain_multiline",
  IDENTIFIER: "identifier",
  SLUG: "slug",
  HOST: "host",
  URL: "url",
  PATH: "path",
  REGISTRY: "registry",
  REGEX: "regex",
  SECRET: "secret",
  CODE: "code",
};

const IDENTIFIER_RE = /^[A-Za-z0-9][A-Za-z0-9._:@/-]*$/;
const SLUG_RE = /^[a-z0-9](?:[a-z0-9-]{0,62}[a-z0-9])?$/;
const IPV4_RE = /^(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}(\/([0-9]|[1-2]\d|3[0-2]))?$/;
const HOST_RE = /^(?=.{1,253}$)([A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?)(\.([A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?))*$/;
const CONTROL_RE = /[\u0000-\u0008\u000b\u000c\u000e-\u001f\u007f]/;
const UNSAFE_MARKUP_RE = /<\s*\/?\s*(script|iframe|object|embed|img|svg|math|link|style|meta|base|form|input|button|textarea|select|option|video|audio|source|body|html)\b|on[a-z]+\s*=|javascript\s*:/i;

export { FIELD_CLASS };

export function sanitizeSingleLineInput(value) {
  return String(value ?? "")
    .replace(/[\u0000-\u001f\u007f]+/g, " ")
    .trim()
    .replace(/\s+/g, " ");
}

export function sanitizeMultilineInput(value) {
  return String(value ?? "")
    .replace(/\r\n/g, "\n")
    .replace(/\r/g, "\n")
    .replace(/[\u0000-\u0008\u000b\u000c\u000e-\u001f\u007f]+/g, " ")
    .trim();
}

export function sanitizeNotificationMessage(value) {
  let text = sanitizeMultilineInput(value)
    .replaceAll("<B>", "<b>")
    .replaceAll("</B>", "</b>")
    .replaceAll("<b>", "\u0000BOLD_OPEN\u0000")
    .replaceAll("</b>", "\u0000BOLD_CLOSE\u0000");
  text = text.replaceAll("<", "").replaceAll(">", "");
  return text
    .replaceAll("\u0000BOLD_OPEN\u0000", "<b>")
    .replaceAll("\u0000BOLD_CLOSE\u0000", "</b>")
    .slice(0, DEFAULT_MAX_LENGTH);
}

export function classifyInputField(field) {
  const key = String(field || "").toLowerCase().trim().replace(/^\[|\]$/g, "");
  if (
    key.includes("password") ||
    key.includes("cipher") ||
    key.includes("secret") ||
    key.includes("token") ||
    key.includes("pem") ||
    key.includes("private_key") ||
    key.includes("backup")
  ) return FIELD_CLASS.SECRET;
  if (
    key.includes("script") ||
    key.includes("code") ||
    key.includes("content") ||
    key.includes("stdout") ||
    key.includes("stderr") ||
    key.includes("command") ||
    key.includes("argument") ||
    key.includes("json") ||
    key.includes("payload") ||
    key === "data" ||
    key === "value_data"
  ) return FIELD_CLASS.CODE;
  if (key.includes("regex")) return FIELD_CLASS.REGEX;
  if (key.includes("registry")) return FIELD_CLASS.REGISTRY;
  if (key === "path" || key.endsWith("_path") || key.includes("directory") || key.includes("folder")) return FIELD_CLASS.PATH;
  if (key.includes("url") || key.includes("uri") || key.includes("endpoint")) return FIELD_CLASS.URL;
  if (key === "hostname" || key.includes("host_name") || key.endsWith("_host") || key.includes("cidr") || key.includes("ip_address")) return FIELD_CLASS.HOST;
  if (key.includes("slug")) return FIELD_CLASS.SLUG;
  if (
    key === "id" ||
    key.endsWith("_id") ||
    key.includes("guid") ||
    key.includes("uuid") ||
    key === "role" ||
    key === "action" ||
    key.endsWith("_action") ||
    key === "platform" ||
    key === "artifact" ||
    key === "branch" ||
    key === "service_mode"
  ) return FIELD_CLASS.IDENTIFIER;
  if (key.includes("description") || key.includes("notes") || key.includes("message") || key.includes("comment")) {
    return FIELD_CLASS.PLAIN_MULTILINE;
  }
  return FIELD_CLASS.PLAIN_SINGLE_LINE;
}

export function validateInputValue(field, value, fieldClass = classifyInputField(field)) {
  const text = String(value ?? "");
  const allowNewlines = [FIELD_CLASS.PLAIN_MULTILINE, FIELD_CLASS.SECRET, FIELD_CLASS.CODE, FIELD_CLASS.REGEX].includes(fieldClass);
  const maxLength = [FIELD_CLASS.SECRET, FIELD_CLASS.CODE].includes(fieldClass)
    ? SECRET_MAX_LENGTH
    : fieldClass === FIELD_CLASS.PATH || fieldClass === FIELD_CLASS.REGISTRY
      ? PATH_MAX_LENGTH
      : fieldClass === FIELD_CLASS.IDENTIFIER
        ? IDENTIFIER_MAX_LENGTH
        : DEFAULT_MAX_LENGTH;

  if (text.length > maxLength) return `${field} exceeds ${maxLength} characters.`;
  if (CONTROL_RE.test(text) || (!allowNewlines && /[\r\n\t]/.test(text))) return `${field} cannot include control characters.`;

  const trimmed = text.trim();
  if (!trimmed) return "";
  if (![FIELD_CLASS.SECRET, FIELD_CLASS.CODE, FIELD_CLASS.PATH, FIELD_CLASS.REGISTRY, FIELD_CLASS.REGEX].includes(fieldClass) && UNSAFE_MARKUP_RE.test(trimmed)) {
    return `${field} cannot include executable markup.`;
  }

  if (fieldClass === FIELD_CLASS.IDENTIFIER && !IDENTIFIER_RE.test(trimmed)) return `${field} must be an identifier, enum, or safe ref.`;
  if (fieldClass === FIELD_CLASS.SLUG && !SLUG_RE.test(trimmed)) return `${field} must be a lowercase slug.`;
  if (fieldClass === FIELD_CLASS.HOST && !isHostInput(trimmed)) return `${field} must be a hostname, IP address, or CIDR.`;
  if (fieldClass === FIELD_CLASS.URL) {
    try {
      const parsed = new URL(trimmed);
      if (!["http:", "https:", "ldap:", "ldaps:"].includes(parsed.protocol)) return `${field} uses an unsupported URL scheme.`;
    } catch {
      return `${field} must be an absolute URL.`;
    }
  }
  if (fieldClass === FIELD_CLASS.REGISTRY && (trimmed.includes("/") || /[<>"|?*]/.test(trimmed))) {
    return `${field} must be a Windows registry path or value name.`;
  }
  if (fieldClass === FIELD_CLASS.REGEX) {
    try {
      new RegExp(trimmed);
    } catch {
      return `${field} must be a valid regular expression.`;
    }
  }
  return "";
}

export function sanitizePayload(value, field = "body", fieldClass = classifyInputField(field)) {
  if (Array.isArray(value)) {
    const errors = [];
    const sanitized = value.map((item, index) => {
      const result = sanitizePayload(item, `${field}[${index}]`, fieldClass);
      errors.push(...result.errors);
      return result.value;
    });
    return { value: sanitized, errors };
  }

  const fileLike = typeof File !== "undefined" && value instanceof File;
  const blobLike = typeof Blob !== "undefined" && value instanceof Blob;
  if (value && typeof value === "object" && !fileLike && !blobLike) {
    const errors = [];
    const sanitized = {};
    for (const [key, child] of Object.entries(value)) {
      const childClass = classifyInputField(key);
      const result = sanitizePayload(child, key, childClass);
      errors.push(...result.errors.map((error) => ({ ...error, field: error.field.includes(".") ? error.field : key })));
      sanitized[key] = result.value;
    }
    return { value: sanitized, errors };
  }

  if (typeof value !== "string") return { value, errors: [] };
  const error = validateInputValue(field, value, fieldClass);
  if (error) return { value, errors: [{ field, message: error }] };

  if ([FIELD_CLASS.PLAIN_SINGLE_LINE, FIELD_CLASS.IDENTIFIER, FIELD_CLASS.SLUG, FIELD_CLASS.HOST, FIELD_CLASS.URL].includes(fieldClass)) {
    return { value: sanitizeSingleLineInput(value), errors: [] };
  }
  if (fieldClass === FIELD_CLASS.PLAIN_MULTILINE) {
    return { value: sanitizeMultilineInput(value), errors: [] };
  }
  return { value, errors: [] };
}

export function validateBorealisFetchRequest(input, init = {}, baseOrigin = globalThis.window?.location?.origin || "http://localhost") {
  const urlText = typeof input === "string" || input instanceof URL ? String(input) : String(input?.url || "");
  let parsed;
  try {
    parsed = new URL(urlText, baseOrigin);
  } catch {
    return { input, init, errors: [] };
  }
  const base = new URL(baseOrigin);
  if (parsed.origin !== base.origin || !parsed.pathname.startsWith("/api/") || parsed.pathname.startsWith("/api/internal/") || parsed.pathname === "/api/server/k3s/operator") {
    return { input, init, errors: [] };
  }

  const errors = [];
  const decodedPath = safeDecodeURIComponent(parsed.pathname);
  if (/[\u0000-\u001f\u007f]/.test(decodedPath) || UNSAFE_MARKUP_RE.test(decodedPath)) {
    errors.push({ field: "path", message: "path cannot include executable markup or control characters." });
  }
  for (const [key, value] of parsed.searchParams.entries()) {
    const error = validateInputValue(`query.${key}`, value, classifyInputField(key));
    if (error) errors.push({ field: `query.${key}`, message: error });
  }

  let nextInit = init;
  const body = init?.body;
  const contentType = headerValue(init?.headers, "content-type");
  if (typeof body === "string" && (contentType.includes("json") || /^[\s\r\n]*[\[{]/.test(body))) {
    try {
      const parsedBody = JSON.parse(body);
      const result = sanitizePayload(parsedBody);
      errors.push(...result.errors);
      if (!errors.length) {
        nextInit = {
          ...init,
          body: JSON.stringify(result.value),
          headers: withJSONContentType(init?.headers),
        };
      }
    } catch {
      errors.push({ field: "body", message: "Request body must be valid JSON." });
    }
  } else if (typeof FormData !== "undefined" && body instanceof FormData) {
    for (const [key, value] of body.entries()) {
      if (typeof File !== "undefined" && value instanceof File) continue;
      const error = validateInputValue(key, value, classifyInputField(key));
      if (error) errors.push({ field: key, message: error });
    }
  }

  return { input, init: nextInit, errors };
}

export function installBorealisInputValidationGuard(win = globalThis.window) {
  if (!win || win.__BorealisInputValidationGuardInstalled || typeof win.fetch !== "function") return;
  const originalFetch = win.fetch.bind(win);
  win.fetch = async (input, init = {}) => {
    const result = validateBorealisFetchRequest(input, init, win.location?.origin || "http://localhost");
    if (result.errors.length) {
      return new Response(JSON.stringify({ error: "validation_failed", errors: result.errors }), {
        status: 400,
        headers: { "Content-Type": "application/json" },
      });
    }
    return originalFetch(result.input, result.init);
  };
  Object.defineProperty(win, "__BorealisInputValidationGuardInstalled", {
    value: true,
    configurable: true,
  });
}

function headerValue(headers, key) {
  if (!headers) return "";
  if (typeof headers.get === "function") return String(headers.get(key) || "").toLowerCase();
  const found = Object.entries(headers).find(([name]) => name.toLowerCase() === key.toLowerCase());
  return String(found?.[1] || "").toLowerCase();
}

function withJSONContentType(headers) {
  if (!headers) return { "Content-Type": "application/json" };
  if (typeof Headers !== "undefined" && headers instanceof Headers) {
    const next = new Headers(headers);
    if (!next.has("Content-Type")) next.set("Content-Type", "application/json");
    return next;
  }
  return { ...headers, "Content-Type": headerValue(headers, "content-type") || "application/json" };
}

function safeDecodeURIComponent(value) {
  try {
    return decodeURIComponent(value);
  } catch {
    return value;
  }
}

function isHostInput(value) {
  if (IPV4_RE.test(value) || HOST_RE.test(value)) return true;
  const parts = value.split("/");
  if (parts.length > 2) return false;
  const [address, prefix] = parts;
  if (prefix !== undefined && (!/^\d+$/.test(prefix) || Number(prefix) > 128)) return false;
  if (!address.includes(":")) return false;
  try {
    new URL(`http://[${address}]`);
    return true;
  } catch {
    return false;
  }
}

File: Data/Engine/Containers/api-backend/cmd/api-backend/input_validation.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"mime"
	"net"
	"net/http"
	"net/netip"
	"net/url"
	"regexp"
	"strings"
	"unicode"
	"unicode/utf8"
)

const (
	publicJSONMapMaxBytes     int64 = 1 << 20
	publicAuthJSONMaxBytes    int64 = 64 << 10
	publicBackupJSONMaxBytes  int64 = 128 << 20
	maxInputPathTextLength          = 4096
	maxInputPlainTextLength         = 4096
	maxInputIdentifierLength        = 256
	maxNotificationTextLength       = 4096
)

var (
	errRequestBodyTooLarge = errors.New("request body too large")
	inputIdentifierRE      = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._:@/\-]*$`)
	inputSlugRE            = regexp.MustCompile(`^[a-z0-9](?:[a-z0-9-]{0,62}[a-z0-9])?$`)
	hostLabelRE            = regexp.MustCompile(`^[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?$`)
	portRE                 = regexp.MustCompile(`^\d{1,5}$`)
	unsafeMarkupRE         = regexp.MustCompile(`(?i)<\s*/?\s*(script|iframe|object|embed|img|svg|math|link|style|meta|base|form|input|button|textarea|select|option|video|audio|source|body|html)\b|on[a-z]+\s*=|javascript\s*:`)
)

type publicValidationError struct {
	Field   string `json:"field"`
	Message string `json:"message"`
}

type inputFieldClass string

const (
	inputClassPlainSingleLine inputFieldClass = "plain_single_line"
	inputClassPlainMultiline  inputFieldClass = "plain_multiline"
	inputClassIdentifier      inputFieldClass = "identifier"
	inputClassSlug            inputFieldClass = "slug"
	inputClassHost            inputFieldClass = "host"
	inputClassURL             inputFieldClass = "url"
	inputClassPath            inputFieldClass = "path"
	inputClassRegistry        inputFieldClass = "registry"
	inputClassRegex           inputFieldClass = "regex"
	inputClassSecret          inputFieldClass = "secret"
	inputClassCode            inputFieldClass = "code"
)

func withPublicInputValidation(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if !publicInputValidationApplies(r) {
			next.ServeHTTP(w, r)
			return
		}
		if errs := validatePublicRequestTarget(r); len(errs) > 0 {
			writePublicValidationErrors(w, errs)
			return
		}
		next.ServeHTTP(w, r)
	})
}

func publicInputValidationApplies(r *http.Request) bool {
	if r == nil || r.URL == nil {
		return false
	}
	path := r.URL.Path
	if !strings.HasPrefix(path, "/api/") {
		return false
	}
	if strings.HasPrefix(path, "/api/internal/") {
		return false
	}
	return path != "/api/server/k3s/operator"
}

func validatePublicRequestTarget(r *http.Request) []publicValidationError {
	errs := make([]publicValidationError, 0)
	if err := validateInputUTF8AndControls("path", r.URL.Path, maxInputPathTextLength, false); err != nil {
		errs = append(errs, publicValidationError{Field: "path", Message: err.Error()})
	} else if err := validateNoUnsafeMarkup("path", r.URL.Path); err != nil {
		errs = append(errs, publicValidationError{Field: "path", Message: err.Error()})
	}
	for key, values := range r.URL.Query() {
		field := "query." + key
		if err := validateInputUTF8AndControls(field, key, maxInputIdentifierLength, false); err != nil {
			errs = append(errs, publicValidationError{Field: field, Message: err.Error()})
			continue
		}
		class := classifyInputField(key)
		for _, value := range values {
			if err := validateInputValue(field, value, class); err != nil {
				errs = append(errs, publicValidationError{Field: field, Message: err.Error()})
			}
		}
	}
	return errs
}

func readJSONMapWithLimit(r *http.Request, limit int64) (map[string]any, error) {
	raw, err := readLimitedRequestBody(r, limit)
	if err != nil {
		return nil, err
	}
	if len(strings.TrimSpace(string(raw))) == 0 {
		return map[string]any{}, nil
	}
	var body map[string]any
	if err := json.Unmarshal(raw, &body); err != nil {
		return nil, err
	}
	if body == nil {
		body = map[string]any{}
	}
	if errs := sanitizeJSONInputMap(body); len(errs) > 0 {
		return nil, publicValidationErrors(errs)
	}
	return body, nil
}

func readLimitedRequestBody(r *http.Request, limit int64) ([]byte, error) {
	if r == nil || r.Body == nil {
		return nil, nil
	}
	if limit <= 0 {
		limit = publicJSONMapMaxBytes
	}
	raw, err := io.ReadAll(io.LimitReader(r.Body, limit+1))
	if err != nil {
		return nil, err
	}
	if int64(len(raw)) > limit {
		return nil, errRequestBodyTooLarge
	}
	return raw, nil
}

func sanitizeJSONInputMap(body map[string]any) []publicValidationError {
	return sanitizeJSONInputValue("", body, inputClassPlainSingleLine)
}

func validateJSONTransportMap(body map[string]any) []publicValidationError {
	return validateJSONTransportValue("", body)
}

func validateJSONTransportValue(path string, value any) []publicValidationError {
	errs := make([]publicValidationError, 0)
	switch typed := value.(type) {
	case map[string]any:
		for key, child := range typed {
			fieldPath := key
			if path != "" {
				fieldPath = path + "." + key
			}
			errs = append(errs, validateJSONTransportValue(fieldPath, child)...)
		}
	case []any:
		for index, child := range typed {
			fieldPath := fmt.Sprintf("%s[%d]", path, index)
			errs = append(errs, validateJSONTransportValue(fieldPath, child)...)
		}
	case string:
		if err := validateInputUTF8AndControls(firstText(path, "body"), typed, 16<<20, true); err != nil {
			errs = append(errs, publicValidationError{Field: firstText(path, "body"), Message: err.Error()})
		}
	}
	return errs
}

func sanitizeJSONInputValue(path string, value any, parentClass inputFieldClass) []publicValidationError {
	errs := make([]publicValidationError, 0)
	switch typed := value.(type) {
	case map[string]any:
		for key, child := range typed {
			fieldPath := key
			if path != "" {
				fieldPath = path + "." + key
			}
			class := classifyInputField(key)
			if text, ok := child.(string); ok {
				sanitized, _ := sanitizeInputValueByClass(text, class)
				if err := validateInputValue(fieldPath, sanitized, class); err != nil {
					errs = append(errs, publicValidationError{Field: firstText(fieldPath, "body"), Message: err.Error()})
					continue
				}
				typed[key] = sanitized
				continue
			}
			childErrs := sanitizeJSONInputValue(fieldPath, child, class)
			if len(childErrs) > 0 {
				errs = append(errs, childErrs...)
			}
		}
	case []any:
		for index, child := range typed {
			fieldPath := fmt.Sprintf("%s[%d]", path, index)
			if text, ok := child.(string); ok {
				sanitized, _ := sanitizeInputValueByClass(text, parentClass)
				if err := validateInputValue(fieldPath, sanitized, parentClass); err != nil {
					errs = append(errs, publicValidationError{Field: firstText(fieldPath, "body"), Message: err.Error()})
					continue
				}
				typed[index] = sanitized
				continue
			}
			childErrs := sanitizeJSONInputValue(fieldPath, child, parentClass)
			if len(childErrs) > 0 {
				errs = append(errs, childErrs...)
			}
		}
	case string:
		sanitized, _ := sanitizeInputValueByClass(typed, parentClass)
		if err := validateInputValue(path, sanitized, parentClass); err != nil {
			errs = append(errs, publicValidationError{Field: firstText(path, "body"), Message: err.Error()})
		}
	}
	return errs
}

func sanitizeInputValueByClass(value string, class inputFieldClass) (string, bool) {
	switch class {
	case inputClassPlainSingleLine, inputClassIdentifier, inputClassSlug, inputClassHost, inputClassURL:
		sanitized := sanitizeSingleLineInput(value)
		return sanitized, sanitized != value
	case inputClassPlainMultiline:
		sanitized := sanitizeMultilineInput(value)
		return sanitized, sanitized != value
	default:
		return value, false
	}
}

func validateInputValue(field string, value string, class inputFieldClass) error {
	switch class {
	case inputClassSecret, inputClassCode:
		return validateInputUTF8AndControls(field, value, 16<<20, true)
	case inputClassPath:
		return validatePathInput(field, value)
	case inputClassRegistry:
		return validateRegistryInput(field, value)
	case inputClassRegex:
		return validateRegexInput(field, value)
	case inputClassURL:
		return validateURLInput(field, value)
	case inputClassHost:
		return validateHostInput(field, value)
	case inputClassSlug:
		return validateSlugInput(field, value)
	case inputClassIdentifier:
		return validateIdentifierInput(field, value)
	case inputClassPlainMultiline:
		if err := validateInputUTF8AndControls(field, value, maxInputPlainTextLength, true); err != nil {
			return err
		}
		return validateNoUnsafeMarkup(field, value)
	default:
		if err := validateInputUTF8AndControls(field, value, maxInputPlainTextLength, false); err != nil {
			return err
		}
		return validateNoUnsafeMarkup(field, value)
	}
}

func validateInputUTF8AndControls(field string, value string, maxLen int, allowNewlines bool) error {
	if !utf8.ValidString(value) {
		return fmt.Errorf("%s must be valid UTF-8", field)
	}
	if maxLen > 0 && len(value) > maxLen {
		return fmt.Errorf("%s exceeds %d bytes", field, maxLen)
	}
	for _, r := range value {
		if r == 0 || r == '\u007f' {
			return fmt.Errorf("%s cannot include control characters", field)
		}
		if unicode.IsControl(r) {
			if allowNewlines && (r == '\n' || r == '\r' || r == '\t') {
				continue
			}
			return fmt.Errorf("%s cannot include control characters", field)
		}
	}
	return nil
}

func validateNoUnsafeMarkup(field string, value string) error {
	if unsafeMarkupRE.MatchString(value) {
		return fmt.Errorf("%s cannot include executable markup", field)
	}
	return nil
}

func validateIdentifierInput(field string, value string) error {
	value = strings.TrimSpace(value)
	if value == "" {
		return nil
	}
	if err := validateInputUTF8AndControls(field, value, maxInputIdentifierLength, false); err != nil {
		return err
	}
	if strings.ContainsAny(value, "<>\\") || !inputIdentifierRE.MatchString(value) {
		return fmt.Errorf("%s must be an identifier, enum, or safe ref", field)
	}
	return nil
}

func validateSlugInput(field string, value string) error {
	value = strings.TrimSpace(value)
	if value == "" {
		return nil
	}
	if err := validateInputUTF8AndControls(field, value, 63, false); err != nil {
		return err
	}
	if !inputSlugRE.MatchString(value) {
		return fmt.Errorf("%s must be a lowercase slug", field)
	}
	return nil
}

func validateHostInput(field string, value string) error {
	value = strings.TrimSpace(value)
	if value == "" {
		return nil
	}
	if err := validateInputUTF8AndControls(field, value, 253, false); err != nil {
		return err
	}
	if strings.ContainsAny(value, "<>\"'`\\ ") {
		return fmt.Errorf("%s must be a hostname, IP address, or CIDR", field)
	}
	if _, err := netip.ParseAddr(value); err == nil {
		return nil
	}
	if _, err := netip.ParsePrefix(value); err == nil {
		return nil
	}
	if strings.ContainsAny(value, "/:") {
		return fmt.Errorf("%s must be a hostname, IP address, or CIDR", field)
	}
	for _, label := range strings.Split(value, ".") {
		if !hostLabelRE.MatchString(label) {
			return fmt.Errorf("%s must be a valid hostname", field)
		}
	}
	return nil
}

func validateURLInput(field string, value string) error {
	value = strings.TrimSpace(value)
	if value == "" {
		return nil
	}
	if err := validateInputUTF8AndControls(field, value, 4096, false); err != nil {
		return err
	}
	parsed, err := url.Parse(value)
	if err != nil || parsed.Scheme == "" || parsed.Host == "" {
		return fmt.Errorf("%s must be an absolute URL", field)
	}
	scheme := strings.ToLower(parsed.Scheme)
	if scheme != "http" && scheme != "https" && scheme != "ldap" && scheme != "ldaps" {
		return fmt.Errorf("%s uses an unsupported URL scheme", field)
	}
	if strings.ContainsAny(parsed.Host, "<>\"'` ") {
		return fmt.Errorf("%s has an invalid host", field)
	}
	return nil
}

func validatePathInput(field string, value string) error {
	value = strings.TrimSpace(value)
	if value == "" {
		return nil
	}
	if err := validateInputUTF8AndControls(field, value, maxInputPathTextLength, false); err != nil {
		return err
	}
	if strings.Contains(value, "\x00") {
		return fmt.Errorf("%s cannot include NUL bytes", field)
	}
	return nil
}

func validateRemoteFileName(field string, value string) error {
	value = strings.TrimSpace(value)
	if value == "" {
		return nil
	}
	if err := validateInputUTF8AndControls(field, value, 255, false); err != nil {
		return err
	}
	if value == "." || value == ".." || strings.Contains(value, "/") || strings.Contains(value, "\\") {
		return fmt.Errorf("%s cannot include path separators or traversal segments", field)
	}
	if strings.ContainsAny(value, `<>:"|?*`) {
		return fmt.Errorf("%s contains characters Windows file systems do not allow", field)
	}
	if strings.HasSuffix(value, ".") || strings.HasSuffix(value, " ") {
		return fmt.Errorf("%s cannot end with a space or period", field)
	}
	return nil
}

func validateRegistryInput(field string, value string) error {
	value = strings.TrimSpace(value)
	if value == "" {
		return nil
	}
	if err := validateInputUTF8AndControls(field, value, maxInputPathTextLength, false); err != nil {
		return err
	}
	if strings.Contains(value, "/") || strings.ContainsAny(value, "<>\"|?*") {
		return fmt.Errorf("%s must be a Windows registry path or value name", field)
	}
	return nil
}

func validateRegexInput(field string, value string) error {
	if value == "" {
		return nil
	}
	if err := validateInputUTF8AndControls(field, value, maxInputPlainTextLength, true); err != nil {
		return err
	}
	if _, err := regexp.Compile(value); err != nil {
		return fmt.Errorf("%s must be a valid regular expression", field)
	}
	return nil
}

func sanitizeSingleLineInput(value string) string {
	value = strings.Map(func(r rune) rune {
		if r == 0 || unicode.IsControl(r) {
			return ' '
		}
		return r
	}, value)
	return strings.Join(strings.Fields(value), " ")
}

func sanitizeMultilineInput(value string) string {
	value = strings.ReplaceAll(value, "\r\n", "\n")
	value = strings.ReplaceAll(value, "\r", "\n")
	value = strings.Map(func(r rune) rune {
		if r == 0 {
			return -1
		}
		if unicode.IsControl(r) && r != '\n' && r != '\t' {
			return ' '
		}
		return r
	}, value)
	return strings.TrimSpace(value)
}

func sanitizeNotificationText(value string) string {
	value = sanitizeMultilineInput(value)
	if value == "" {
		return ""
	}
	value = strings.ReplaceAll(value, "<B>", "<b>")
	value = strings.ReplaceAll(value, "</B>", "</b>")
	value = strings.ReplaceAll(value, "<b>", "\x00BOLD_OPEN\x00")
	value = strings.ReplaceAll(value, "</b>", "\x00BOLD_CLOSE\x00")
	value = strings.ReplaceAll(value, "<", "")
	value = strings.ReplaceAll(value, ">", "")
	value = strings.ReplaceAll(value, "\x00BOLD_OPEN\x00", "<b>")
	value = strings.ReplaceAll(value, "\x00BOLD_CLOSE\x00", "</b>")
	if len(value) > maxNotificationTextLength {
		value = value[:maxNotificationTextLength]
	}
	return value
}

func classifyInputField(field string) inputFieldClass {
	key := strings.ToLower(strings.TrimSpace(field))
	key = strings.Trim(key, "[]")
	switch {
	case strings.Contains(key, "password"), strings.Contains(key, "cipher"), strings.Contains(key, "secret"),
		strings.Contains(key, "token"), strings.Contains(key, "pem"), strings.Contains(key, "private_key"),
		strings.Contains(key, "backup"), key == "invite_bundle":
		return inputClassSecret
	case strings.Contains(key, "script"), strings.Contains(key, "code"), strings.Contains(key, "content"),
		strings.Contains(key, "stdout"), strings.Contains(key, "stderr"), strings.Contains(key, "command"),
		strings.Contains(key, "argument"), strings.Contains(key, "json"), strings.Contains(key, "payload"),
		key == "data" || key == "value_data":
		return inputClassCode
	case strings.Contains(key, "regex"):
		return inputClassRegex
	case strings.Contains(key, "registry"):
		return inputClassRegistry
	case key == "path" || strings.HasSuffix(key, "_path") || strings.Contains(key, "directory") || strings.Contains(key, "folder"):
		return inputClassPath
	case strings.Contains(key, "url") || strings.Contains(key, "uri") || strings.Contains(key, "endpoint"):
		return inputClassURL
	case key == "hostname" || strings.Contains(key, "host_name") || strings.HasSuffix(key, "_host") || strings.Contains(key, "cidr") || strings.Contains(key, "ip_address"):
		return inputClassHost
	case strings.Contains(key, "slug"):
		return inputClassSlug
	case key == "id" || strings.HasSuffix(key, "_id") || strings.Contains(key, "guid") || strings.Contains(key, "uuid") ||
		key == "role" || key == "action" || strings.HasSuffix(key, "_action") || key == "platform" ||
		key == "artifact" || key == "branch" || key == "service_mode":
		return inputClassIdentifier
	case strings.Contains(key, "description") || strings.Contains(key, "notes") || strings.Contains(key, "message") || strings.Contains(key, "comment"):
		return inputClassPlainMultiline
	default:
		return inputClassPlainSingleLine
	}
}

type publicValidationErrors []publicValidationError

func (e publicValidationErrors) Error() string {
	return "validation_failed"
}

func asPublicValidationErrors(err error) ([]publicValidationError, bool) {
	var errs publicValidationErrors
	if errors.As(err, &errs) {
		return []publicValidationError(errs), true
	}
	return nil, false
}

func writePublicValidationErrors(w http.ResponseWriter, errs []publicValidationError) {
	if len(errs) == 0 {
		errs = []publicValidationError{{Field: "body", Message: "Request input is invalid."}}
	}
	writeJSON(w, http.StatusBadRequest, map[string]any{
		"error":  "validation_failed",
		"errors": errs,
	})
}

func publicValidationErrorPayload(err error, fallbackError string) map[string]any {
	if errors.Is(err, errRequestBodyTooLarge) {
		return map[string]any{"error": "request_body_too_large"}
	}
	if errs, ok := asPublicValidationErrors(err); ok {
		if len(errs) == 0 {
			errs = []publicValidationError{{Field: "body", Message: "Request input is invalid."}}
		}
		return map[string]any{"error": "validation_failed", "errors": errs}
	}
	return map[string]any{"error": firstText(fallbackError, "invalid_json")}
}

func invalidJSONOrValidation(w http.ResponseWriter, err error) {
	if errors.Is(err, errRequestBodyTooLarge) {
		writeJSON(w, http.StatusBadRequest, map[string]any{"error": "request_body_too_large"})
		return
	}
	if errs, ok := asPublicValidationErrors(err); ok {
		writePublicValidationErrors(w, errs)
		return
	}
	writeJSON(w, http.StatusBadRequest, map[string]any{"error": "invalid_json"})
}

func requestHasJSONBody(r *http.Request) bool {
	if r == nil || r.Body == nil {
		return false
	}
	mediaType, _, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
	if err != nil {
		return strings.Contains(strings.ToLower(r.Header.Get("Content-Type")), "json")
	}
	return mediaType == "application/json" || strings.HasSuffix(mediaType, "+json")
}

func validHostPort(host string, port string) bool {
	if strings.TrimSpace(host) == "" {
		return false
	}
	if ip := net.ParseIP(host); ip != nil {
		return true
	}
	return validateHostInput("host", host) == nil && (port == "" || portRE.MatchString(port))
}
